generators/add-licence-headers/index.js

```diff
@@ -0,0 +1,192 @@
+      this.licenceText = this.fs.read(this.licencePath);
+    } catch (e) {
+      return this._abort(e);
+    }
+  }
+
+  _generateComments() {
+    // get maximum line length by looping through all lines, returning the line length and passing them to Math.max
+    const lineArray = this.licenceText.split('\n');
+    const maxLineLength = Math.max(...lineArray.map((line) => line.length));
+
+    const align = (string, threshold, timesSpaces) => {
+      if (!timesSpaces) {
+        timesSpaces = threshold;
+      }
+
+      if (string.length > threshold) {
+        return ' '.repeat(timesSpaces);
+      }
+      return '';
+    };
+
+    Object.keys(comments).forEach((fileType) => {
+      const commentStyle = comments[fileType];
+      let comment = commentStyle.begin + commentStyle.body.repeat(maxLineLength) + '\n';
+      lineArray.forEach((line) => {
+        // only add a space character when the opening comment contains more than one charcters (e.g. to align the asterisks in .ts files vs. aligning the hashes in .py files)
+        comment += align(commentStyle.begin, commentStyle.aligningSpaces);
+        comment += `${commentStyle.body}${align(line, 0, 1) + line}\n`;
+      });
+
+      comment += align(commentStyle.begin, commentStyle.aligningSpaces);
+      comment += `${commentStyle.body.repeat(maxLineLength)}${commentStyle.end}`;
+      this.comments[fileType] = comment;
+    });
+  }
+
+  _getSourceFolders() {
+    const backendSourcePath = path.basename(this.destinationPath());
+    return ['src', backendSourcePath];
+  }
+
+  _abort(msg) {
+    return Promise.reject(msg ? msg : 'Step Failed: Aborting');
+  }
+
+  _addComments(path, fileExtension) {
+    const folderContents = glob.sync(`**/*.${fileExtension}`, {cwd: path});
+
+    if (folderContents.length === 0) {
+      return;
+    }
+
+    folderContents.forEach((file) => {
+      const filePath = path + '/' + file;
+      let fileContents = this.fs.read(filePath);
+
+      // TODO: override any comment if the file starts with one (e.g. when the licence changes)
+      // whenever a file starts with our licence header skip for now
+      if (fileContents.startsWith(this.comments[fileExtension])) {
+        return;
+      }
+
+      fileContents = this._findAndRemoveHeader(fileContents, fileExtension);
+
+      const newContents = this.comments[fileExtension] + os.EOL + os.EOL + fileContents;
+      this.fs.write(filePath, newContents);
+    });
+  }
+
+  /**
+   * remove a multiline comment at the beginning of a file (e.g. licence header) if it exists
+   * @param fileContents
+   * @param fileExtension
+   * @returns {string}
+   * @private
+   */
+  _findAndRemoveHeader(fileContents, fileExtension) {
+    if (fileContents.startsWith(comments[fileExtension].begin)) {
+      const linesArray = fileContents.split(os.EOL);
+      let line = linesArray.shift();
+      while (line.startsWith(comments[fileExtension].begin) || line.indexOf(comments[fileExtension].body) > -1) {
```

Comment:
```
line.indexOf(comments[fileExtension].body) > -1
```
is very risky in this current format, e.g for typescript body=*

and thus will match
```
import * as d3 from 'd3';
```

shouldn't it start with the * but ignoring leading whitespaces?

Reply:
in my tests it worked, but I think only because there was a newline between the comment headers and the imports, therefore the line didn't start with the comment beginning and the comment body.

But your're right, I didn't think of the case that there could be the newline missing and I think in this case the more would be overridden.

I changed it to this condition:
`while (line.startsWith(commentConfig.begin) || line.indexOf(commentConfig.body) === comments[fileExtension].aligningSpaces)`

The reason is that the comment body should be at index 0 for Python files and for SCSS and TS at index 1 (because they're aligned with spaces).

An alternative would be (which would be less to read):
`while (line.startsWith(commentConfig.begin) || line.trim().startsWith(commentConfig.body)`

I think then the condition would be safe. What do you think? Which one do you prefer?